src/libcore/sync/atomic.rs

```diff
@@ -1166,6 +1166,216 @@ impl<T> AtomicPtr<T> {
             }
         }
     }
+
+    /// Adds to the current pointer, returning the previous pointer.
+    ///
+    /// Unlike other pointer additions, `fetch_add` increments directly by the provided value,
+    /// rather than interpreting it as a multiple of `size_of<T>`.
+    ///
+    /// This operation wraps around on overflow.
+    ///
+    /// `fetch_add` takes an [`Ordering`] argument which describes the memory ordering
+    /// of this operation. All ordering modes are possible. Note that using
+    /// [`Acquire`] makes the store part of this operation [`Relaxed`], and
+    /// using [`Release`] makes the load part [`Relaxed`].
+    ///
+    /// [`Ordering`]: enum.Ordering.html
+    /// [`Relaxed`]: enum.Ordering.html#variant.Relaxed
+    /// [`Release`]: enum.Ordering.html#variant.Release
+    /// [`Acquire`]: enum.Ordering.html#variant.Acquire
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(atomic_ptr_fetch_op)]
+    /// use std::sync::atomic::{AtomicPtr, Ordering};
+    ///
+    /// let foo = AtomicPtr::new(0 as *mut ());
+    /// assert_eq!(foo.fetch_add(10, Ordering::SeqCst), 0 as *mut _);
+    /// assert_eq!(foo.load(Ordering::SeqCst), 10 as *mut _);
+    /// ```
+    #[inline]
+    #[cfg(target_has_atomic = "ptr")]
+    #[unstable(feature = "atomic_ptr_fetch_op", issue = "none")]
+    pub fn fetch_add(&self, val: usize, order: Ordering) -> *mut T {
+        // SAFETY: data races are prevented by atomic intrinsics.
```

Comment:
This doesn't seem to have anything to do with the actual sources of unsafety in the block.

Reply:
Ah, that safety comment mirrors that of the other atomic ops, and reflects the reason the `atomic_add` unsafety is safe. Are you referring to the additional unsafety in transmuting from a `usize` to a `*mut T`?

Reply:
The unsafe calls in that block are 1. transmuting, and 2. calling the unsafe intrinsic that takes a raw pointer.